ICQ Slashing

riley-stride opened the pull request:
## Summary
When validators on host zones are slashed, we need to update our records of their balances accordingly, or users would redeem at too favorable an exchange rate! Here's we expose a stakeibc message update-delegation  that allows for querying a particular host zone validator pair using ICQ to update our record of that validator's delegation. We rate-limit using a safety bounds : the queried result cannot be -10% or less of the unslashed value.

To avoid race conditions between ICAs and these ICQs, we divvy up each epoch into two portions (currently split 80/20). Stakeibc now has a param that determines the split point. 
- In the first portion, we allow ICAs (they time out at the split point). 
- In the second, we allow the ICQs implemented in this PR (they time out at the end of the epoch).  Any ICQs to delegation amounts submitted before the split point will fail.


Note that to query a (del,val) delegation on a host zone, two ICQs are required! This is due to a limitation in ICQ: we can only query (with proofs) data that exists at a single key in the store (see gRPC query limitations in the sdk). 

How do the queries work? Validators have two attributes: shares and tokens. Number of shares do not change, but the internal per-validator exchRate shares:tokens does change as rewards are reinvested or slashing occurs. We need two queries - one to get the internal exchRate and a second to get the number of shares on the (delegator,validator) pair. Armed with these two values, we obtain `num_tokens=exchRate*numShares` (note: it is a bit strange that no query exists get the tokens directly.) Our approach is to:
1. Submit a query to a validator to get the exchRate. In the CB, store that exchRate on the hostZone.validator.TokensFromShares field, then kick off a daisy chained ICQ to the (del,val) pair
2. In the (del,val) pair ICQ's callback, convert the returned shares to tokens, and if they differ from num tokens on the record, update the records (subject to a safety constraints that the slash was not GT 10%). Also update the weight of this validator in line with the slash, as a penalty.

## Description
- Add timeouts to ICA calls in msg_server_submit_tx.go and a corresponsing param to split the epoch up in params.go
- Scaffold a `MsgUpdateValidatorSharesExchRate` message that user can call to query delegations from delegationAcct to a particular validator on a host zone
     - This fn executes an ICQ request to get a validator's [`tokensFromShares()`](https://github.com/cosmos/cosmos-sdk/blob/0943a702155f359fa8a7ed6e7b441f65e17f44d8/x/staking/types/validator.go#L315) exch rate
- Add callbacks to handle "validator" and "delegation" queries. The "validator" query comes first and daisy chains the "delegation" query after storing val's internal exchRate.
- Add some helpers in stakeibc keeper for grabbing end-of-epoch unix time


## HOW TO TEST
### Slashing ICQ
Checkout commit `a929879`, which contains testing settings that make testing the slashing ICQ case convenient (e.g. stringent jailing params, low stake on the validator that will be jailed, etc)

To get the chain started  and liquid stake (you can kill the integration tests once it has liquid staked): 
```make init-local build=s; sh scripts-local/test-utils/1.sh; make test-integration```
 1. Slash validator 2 on gaia
- Find the gaia2 validator's process with `ps aux | grep gaia2/gaiad`
- Kill that process with `kill -9 <PID>`
- Wait 10 blocks and check to make sure the val was jailed (on the explorer or with `gaiad q staking validators` and check the status is JAILED)
    - you can alternatively use `gaiad q distribution slashes cosmosvaloper133lfs9gcpxqj6er3kx605e3v9lqp2pg5syhvsz 0 100000`
3.  Suppose we saw the slash and wanted to update records to reflect it. Submit: `tx stakeibc update-delegation uatom cosmosvaloper133lfs9gcpxqj6er3kx605e3v9lqp2pg5syhvsz --from admin -y`
    - it's possible you submitted the update outside the ICQ buffer window (the last 20% of the epoch); the logs will tell you. Try again if so.
5.  ctrl+F the stride logs for `ICQ'd delAmt mismatch`, which will show you the detected slash amount. It should match gaia's config.toml slash_fraction_downtime param, 0.1
6. Check accounts.log to verify that the validator's delegatedAmt and the hostZone's stakedBal decremented as well! The exchange rate will update the next epoch

### Other
- Unit tests ✓
- Integration tests ✓
- Firehose tests ✓ (except `isClaimable`, WIP) 

## FUTURE ISSUES
1. The icq relayer dies sporadically with an error that says the validator set hash has changed. I suspect this is because we don't update the LC before issuing another query after the val set changed. 
    - For now, naively, we'll have multiple ICQ relayers running so we could be careful to wait until an ICQ has relayed before submitting a query-delegations msg, and if we do by accident, rely on relayers having hot restart scripts.
2. We could probably name `QueryDelegation` and `QueryDelegationDaisyChain` more clearly
3. We need ICQ TTLs

asalzmann commented on x/stakeibc/keeper/callbacks.go:
```diff
@@ -156,3 +160,112 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
+			// note: truncateInt per https://github.com/cosmos/cosmos-sdk/blob/cb31043d35bad90c4daa923bb109f38fd092feda/x/staking/types/validator.go#L431
+			qNumTokens := qdel.Shares.Mul(v.TokensFromShares).TruncateInt()
+			k.Logger(ctx).Info(fmt.Sprintf("DelegationCallback: zone %s validator %s prevNtokens %v qNumTokens %v", zone.ChainId, v.Address, v.DelegationAmt, qNumTokens))
+			if qNumTokens.Uint64() != v.DelegationAmt {
```
Below we're using `slashAmt` - can this code only be executed when a slash occurs? Should we check that `qNumTokens.Uint64() < v.DelegationAmt`?

nit: in the check, should we add some precision factor? e.g. `qNumTokens.Uint64() < (v.DelegationAmt - epsilon)`

riley-stride replied:
Good catch on the LT check. On precision, I don't think so. These are already integers and any epsilon GT 0 would cause problems if tokens appreciate to high USD values (e.g. you don't want to abort the slash update if the slash is 1 token when that token is worth USD1K, for example)

asalzmann commented on x/stakeibc/keeper/callbacks.go:
```diff
@@ -156,3 +160,112 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
+			qNumTokens := qdel.Shares.Mul(v.TokensFromShares).TruncateInt()
+			k.Logger(ctx).Info(fmt.Sprintf("DelegationCallback: zone %s validator %s prevNtokens %v qNumTokens %v", zone.ChainId, v.Address, v.DelegationAmt, qNumTokens))
+			if qNumTokens.Uint64() != v.DelegationAmt {
+				// TODO add some safety checks here
```
What's this TODO?

riley-stride replied:
add some safety checks here (e.g. we could query the slashing module to confirm the decr in tokens was due to slash). I think until ICQ is more robust we should opt to not issue this check, as it would require a third query.

asalzmann reviewed (commented):
Looking good, thanks for taking this! Left a bunch of nit comments and have a few conceptual questions
- `TokensFromShares` is a bit confusing - what's it used for?
- Why do we need a second buffer for ICQs? My thought was that we'd issue ICQs in the existing buffer (last X% of the epoch) and add a TTL to the ICQs such that they time out before the next epoch
- Would it make sense to pull the `IsWithinICQEpochWindow` to the ICQ call sites?

asalzmann commented on x/stakeibc/keeper/callbacks.go:
```diff
@@ -156,3 +160,112 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
+				// TODO add some safety checks here
+
+				// update our records of the total stakedbal and of the validator's delegation amt
+				slashAmt := v.DelegationAmt - qNumTokens.Uint64()
```
note: there's an assumption here that any decrease in delegation amount is a slash - let's add a comment to make note of this

asalzmann commented on x/stakeibc/keeper/callbacks.go:
```diff
@@ -156,3 +160,112 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
+				k.Logger(ctx).Info(fmt.Sprintf("ICQ'd delAmt mismatch zone %s validator %s delegator %s records was %d icq shows %d slashAmt %d slashPct %d... UPDATING!",
+					zone.ChainId, v.Address, qdel.DelegatorAddress, v.DelegationAmt, qNumTokens, slashAmt, slashPct))
+				k.Logger(ctx).Info(fmt.Sprintf("ICQ'd: slashPct: %v", slashPct))
+				if slashPct.GT(sdk.NewDec(10).Quo(sdk.NewDec(100))) {
```
nit: we should move this type of logic to a rate limiting module, let's add a TODO to move this to that module (once we have it)

riley-stride commented on scripts-local/test-utils/1.sh:
```diff
@@ -13,13 +13,13 @@ $STRIDE_CMD q bank balances $STRIDE_ADDRESS
-$GAIA_CMD tx bank send gval1 $GAIA_VAL_2_ADDR 10000uatom --chain-id $GAIA_CHAIN --keyring-backend test -y
-CSLEEP 10
-$GAIA_CMD tx bank send gval1 $GAIA_VAL_3_ADDR 10000uatom --chain-id $GAIA_CHAIN --keyring-backend test -y
+# CSLEEP 10
```
validators are already funded in the startup script so I removed this bc it's redundant and slows down the dev cycle

asalzmann commented:
following up on this - I know I left a bunch of comments, lmk if you want to chat through it live!

riley-stride commented on proto/stakeibc/params.proto:
```diff
@@ -26,5 +26,6 @@ message Params {
   uint64 validator_rebalancing_threshold = 8;
   uint64 ica_timeout_nanos = 9;
   uint64 buffer_size = 10;
+  uint64 icq_buffer_size = 12;
```
just using buffer now

riley-stride commented on x/stakeibc/keeper/callbacks.go:
```diff
@@ -156,3 +160,112 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
+	k.Logger(ctx).Info(fmt.Sprintf("ValidatorCallback: zone %v qval %v", zone.ChainId, qval))
+
+	// ensure ICQ can be issued now! else fail the callback
+	valid, err := k.IsWithinICQEpochWindow(ctx)
```
Haven't seen TTLs work yet, I think safer to do this check (it's pretty inexpensive) on both ends for now

asalzmann replied:
Is it possible this leads to unexpected behavior without a TTL? E.g.
```
// 1. QueryValidatorExchangeRate
// Second half of Epoch 1
isWithinBufferWindow(query) == true
// 2. ValidatorCallback
// Second half of Epoch 1
isWithinBufferWindow(query) == true
// 3. QueryDelegationDaisyChain
// time passes
// Second half of Epoch 2
isWithinBufferWindow(query) == true
// 4. DelegationCallback
// Second half of Epoch 2
isWithinBufferWindow(query) == true
```
I think what we want here is for ICQs / ICAs not to overlap, so they'd have to execute in the _same_ window

riley-stride replied:
Agreed if we cross the end of an epoch we could see this behavior [](edited-with-Axolo)

asalzmann replied:
What ticket is tracking adding TTLs to the queries?

riley-stride commented on x/stakeibc/handler.go:
```diff
@@ -52,6 +52,9 @@ func NewHandler(k keeper.Keeper) sdk.Handler {
 		case *types.MsgRestoreInterchainAccount:
 			res, err := msgServer.RestoreInterchainAccount(sdk.WrapSDKContext(ctx), msg)
 			return sdk.WrapServiceResult(ctx, res, err)
+		case *types.MsgQueryDelegation:
```
Good flag. Updated to `UpdateValidatorSharesExchRate` which I think is most precise

riley-stride commented on x/stakeibc/keeper/msg_server_query_delegation.go:
```diff
@@ -0,0 +1,47 @@
+func (k msgServer) QueryDelegation(goCtx context.Context, msg *types.MsgQueryDelegation) (*types.MsgQueryDelegationResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	hostZone, err := k.GetHostZoneFromHostDenom(ctx, msg.Hostdenom)
```
yes good idea, done

sampocs reviewed (commented):
Nicely done! Couple nit comments below.

Also I'm wondering if there's a better way to organize/name these functions such that the flow is a little more clear. Here's one suggestion - no pressure, but lmk your thoughts!

* Rename MsgQueryDelegation to something that captures the _why_ more than the just the _what_, since this is user facing (maybe something like `MsgQuerySlashedValidator`)
* Move body of this function to a new function in `msg_server_submit_tx.go` so the two functions in the daisy chain are next to each other (so `MsgQuerySlashedValidator` basically just calls this new function)
* Then in `msg_server_submit_tx.go`, the two functions that make up the daisy chain could be `QueryValidatorExchangeRate` and `QueryValidatorShares` 
* Finally, in the callbacks, you'd have `ValidatorExchangeRateCallback` and `ValidatorSharesCallback` 
* Also this last point is VERY nit, but since the exchange rate query happens before the shares one, it might be a little easier to follow if the exchange rate one is defined first in `callbacks.go`

Feel free to spin off any of that or leave it if you don't think it's too nit!

sampocs commented on x/stakeibc/keeper/callbacks.go:
```diff
@@ -156,3 +160,113 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
+	for i, v := range zone.Validators {
+		if v.Address == qval.OperatorAddress {
+			// converting 1.0 gives us the exchange rate to later use in the next CB
+			v.TokensFromShares = qval.TokensFromShares(sdk.NewDec(1.0))
```
can you elaborate on why this is 1.0?

riley-stride replied:
`TokensFromShares` is a fn that multiplies its input by a constant (which is derived within the `TokensFromShares ` fn itself). To get that constant out of the fn, we can pass in `1.0`. I prefer this approach to copying over the internal logic of `TokensFromShares`, because with this approach, even if the internals of `TokensFromShares`, change, we're safe (bc we use the cosmos sdk core module version)

riley-stride commented on x/stakeibc/keeper/msg_server_update_validator_shares_exch_rate.go:
```diff
@@ -0,0 +1,55 @@
+	if err != nil {
+		return nil, err
+	} else if !valid {
+		return nil, sdkerrors.Wrapf(types.ErrOutsideIcqWindow, "outside the buffer time during which ICQs are allowed (%s)", msg.ChainId)
```
yeah they will have to retry if they submit it outside the buffer window, but the fn is admin for now so not a worry

riley-stride commented on x/stakeibc/keeper/callbacks.go:
```diff
@@ -156,3 +160,112 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
+}
+
+// ValidatorCallback is a callback handler for validator queries.
+func ValidatorCallback(k Keeper, ctx sdk.Context, args []byte, query icqtypes.Query) error {
```
I agree in theory on limiting daisy chain lengths. This one in particular only ever has two steps (1) validator query (2) delegation query. 

A simple limit would be a `queriesRemaining` int var that decrements with each CB and resets at the top of the epoch

riley-stride commented on x/stakeibc/keeper/keeper.go:
```diff
@@ -136,3 +140,44 @@ func (k Keeper) GetConnectionId(ctx sdk.Context, portId string) (string, error)
 }
+
+// helper to get the time of the next epoch end
+func (k Keeper) GetEpochEndTime(ctx sdk.Context) (uint64, error) {
```
removed

riley-stride commented on x/stakeibc/types/message_query_delegation.go:
```diff
@@ -0,0 +1,70 @@
+	if err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid creator address (%s)", err)
+	}
+	if err := utils.ValidateAdminAddress(msg.Creator); err != nil {
```
yes we prob want a `trustedEntity` admin list. tracking on JIRA

riley-stride commented on x/stakeibc/types/message_query_delegation.go:
```diff
@@ -0,0 +1,70 @@
+	if len(msg.Valoper) == 0 {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "valoper is required")
+	}
+	if !strings.Contains(msg.Valoper, "valoper") {
```
done

asalzmann replied:
I'm not seeing this change (might be missing it)

riley-stride replied:
The check is in `QueryValidatorExchangeRate` -- it requires getting a hostZone from store, which can't be done in ValidateBasic [](edited-with-Axolo)

riley-stride replied:
https://github.com/Stride-Labs/stride/pull/162/files#diff-927e36914bd9915629c942ba7dd456d78a5fe6d4fd8594371294a699302df5b0R387

riley-stride commented on x/stakeibc/keeper/msg_server_query_delegation.go:
```diff
@@ -0,0 +1,47 @@
+		return nil, sdkerrors.Wrapf(types.ErrInvalidHostZone, "no host zone found for denom (%s)", msg.Hostdenom)
+	}
+
+	_, valAddr, _ := bech32.DecodeAndConvert(msg.Valoper)
```
done

asalzmann replied:
I'm not seeing this change (might be missing it)

riley-stride replied:
It's possible these comments got decoupled from the latest bc file names changed. [Here's the error check](https://github.com/Stride-Labs/stride/pull/162/files#diff-927e36914bd9915629c942ba7dd456d78a5fe6d4fd8594371294a699302df5b0R392)

riley-stride commented on x/stakeibc/keeper/msg_server_query_delegation.go:
```diff
@@ -0,0 +1,47 @@
+	_, valAddr, _ := bech32.DecodeAndConvert(msg.Valoper)
+	data := stakingtypes.GetValidatorKey(valAddr)
+
+	key := "store/staking/key"
```
done

riley-stride commented on x/stakeibc/keeper/msg_server_query_delegation.go:
```diff
@@ -0,0 +1,47 @@
+)
+
+// todo rename for clarity (this is the validator query : 1st step of daisy chain)
+func (k msgServer) QueryDelegation(goCtx context.Context, msg *types.MsgQueryDelegation) (*types.MsgQueryDelegationResponse, error) {
```
that is correct

riley-stride commented on x/stakeibc/keeper/callbacks.go:
```diff
@@ -156,3 +160,112 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
+	if !found {
+		return fmt.Errorf("no registered zone for chain id: %s", query.GetChainId())
+	}
+	qval := stakingtypes.Validator{}
```
done

riley-stride commented on x/stakeibc/keeper/callbacks.go:
```diff
@@ -156,3 +160,112 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
+	}
+
+	// set the validator's conversion rate
+	for i, v := range zone.Validators {
```
done

riley-stride commented on x/stakeibc/keeper/callbacks.go:
```diff
@@ -156,3 +160,112 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
+	// armed with the exch rate, we can now query the (val,del) delegation
+	err = k.QueryDelegationDaisyChain(ctx, zone, qval.OperatorAddress)
+	if err != nil {
+		k.Logger(ctx).Error("ValidatorCallback: failed to query delegation", "zone", zone.ChainId, "err", err)
```
done

riley-stride commented on x/stakeibc/keeper/callbacks.go:
```diff
@@ -156,3 +160,112 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
+	qdel := stakingtypes.Delegation{}
+	err = k.cdc.Unmarshal(args, &qdel)
+	if err != nil {
+		k.Logger(ctx).Error("unable to unmarshal qdel info for zone", "zone", zone.ChainId, "err", err)
```
done

riley-stride commented:
> Nicely done! Couple nit comments below.
> 
> Also I'm wondering if there's a better way to organize/name these functions such that the flow is a little more clear. Here's one suggestion - no pressure, but lmk your thoughts!

Definitely, good callout, cleaning the names up!

> * Rename MsgQueryDelegation to something that captures the _why_ more than the just the _what_, since this is user facing (maybe something like `MsgQuerySlashedValidator`)

Updated, see new name

> * Move body of this function to a new function in `msg_server_submit_tx.go` so the two functions in the daisy chain are next to each other (so `MsgQuerySlashedValidator` basically just calls this new function)
> * Then in `msg_server_submit_tx.go`, the two functions that make up the daisy chain could be `QueryValidatorExchangeRate` and `QueryValidatorShares`

`UpdateValidatorSharesExchRate` has a msg_server and `UpdateDelegationsIcq` does not so slight preference to keep the former in its own file per convention and the latter in submit_tx . Wdyt?


> * Finally, in the callbacks, you'd have `ValidatorExchangeRateCallback` and `ValidatorSharesCallback`

`ValidatorExchangeRateCallback` makes sense, the other I'm calling `DelegatorSharesCallback` as its slighly more accurate (2nd query is to a delegation, not a validator). Good idea.

> * Also this last point is VERY nit, but since the exchange rate query happens before the shares one, it might be a little easier to follow if the exchange rate one is defined first in `callbacks.go`

Good idea, updated

> Feel free to spin off any of that or leave it if you don't think it's too nit!

riley-stride commented on x/stakeibc/keeper/msg_server_submit_tx.go:
```diff
@@ -202,9 +202,16 @@ func (k Keeper) SubmitTxsEpoch(ctx sdk.Context, connectionId string, msgs []sdk.
+	icqBufferSize := cast.ToInt64(k.GetParam(ctx, types.KeyIcqBufferSize))
 	BUFFER := epochTracker.Duration / bufferSize
-	timeoutNanos := epochTracker.NextEpochStartTime - BUFFER
+	ICQBUFFER := epochTracker.Duration / icqBufferSize
```
done

riley-stride commented on scripts-local/test-utils/1.sh:
```diff
@@ -13,13 +13,13 @@ $STRIDE_CMD q bank balances $STRIDE_ADDRESS
-CSLEEP 10
-$GAIA_CMD tx bank send gval1 $GAIA_VAL_3_ADDR 10000uatom --chain-id $GAIA_CHAIN --keyring-backend test -y
+# CSLEEP 10
+# $GAIA_CMD tx bank send gval1 $GAIA_VAL_2_ADDR 10000uatom --chain-id $GAIA_CHAIN --keyring-backend test -y
```
done

riley-stride commented on x/stakeibc/keeper/callbacks.go:
```diff
@@ -45,7 +47,9 @@ func (c Callbacks) AddCallback(id string, fn interface{}) types.QueryCallbacks {
 
 func (c Callbacks) RegisterCallbacks() types.QueryCallbacks {
-	a := c.AddCallback("withdrawalbalance", Callback(WithdrawalBalanceCallback))
+	a := c.AddCallback("withdrawalbalance", Callback(WithdrawalBalanceCallback)).
+		AddCallback("delegation", Callback(DelegationCallback)).
+		AddCallback("validator", Callback(ValidatorCallback))
 	return a.(Callbacks)
```
good suggestion, added independently

riley-stride commented on x/stakeibc/keeper/callbacks.go:
```diff
@@ -156,3 +160,113 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
+				// update our records of the total stakedbal and of the validator's delegation amt
+				// NOTE:  we assume any decrease in delegation amt that's not tracked via records is a slash
+				slashAmt := v.DelegationAmt - qNumTokens.Uint64()
+				slashPct := sdk.NewDec(cast.ToInt64(slashAmt)).Quo(sdk.NewDec(cast.ToInt64(v.DelegationAmt)))
```
done

riley-stride commented on x/stakeibc/keeper/msg_server_update_validator_shares_exch_rate.go:
```diff
@@ -0,0 +1,55 @@
+		return nil, sdkerrors.Wrapf(types.ErrInvalidHostZone, "no host zone found for denom (%s)", msg.ChainId)
+	}
+
+	_, valAddr, _ := bech32.DecodeAndConvert(msg.Valoper)
```
done

riley-stride commented on x/stakeibc/keeper/msg_server_submit_tx.go:
```diff
@@ -247,6 +249,10 @@ func (k Keeper) SubmitTxsEpoch(
 	bufferSize := cast.ToInt64(k.GetParam(ctx, types.KeyBufferSize))
 	BUFFER := epochTracker.Duration / bufferSize
```
nah, fixed

sampocs commented:
> UpdateValidatorSharesExchRate has a msg_server and UpdateDelegationsIcq does not so slight preference to keep the former in its own file per convention and the latter in submit_tx . Wdyt?

So just to clarify, I meant something like this:

```
# msg_server_query_slashed_validator.go
func QuerySlashedValidator () {
    k.QueryValidatorExchangeRate()
}

# msg_server_submit_tx.go
func QueryValidatorExchangeRate() {

}
func QueryValidatorShares() {

}
```
Idea being that QuerySlashedValidator is the user facing function. And QueryValidatorExchangeRate and QueryValidatorShares are the more backend functions that are triggered. 

Is that how you interpreted my message?

sampocs commented:
I've also kind of interpretted msg_server_submit_tx as a hub for submitting icas/icaq. But I might be off on that

riley-stride commented on x/stakeibc/keeper/keeper.go:
```diff
@@ -140,3 +143,44 @@ func (k Keeper) GetConnectionId(ctx sdk.Context, portId string) (string, error)
+	if err != nil {
+		return false, err
+	}
+	bufferSize := cast.ToInt64(k.GetParam(ctx, types.KeyBufferSize))
```
done

riley-stride commented:
> > UpdateValidatorSharesExchRate has a msg_server and UpdateDelegationsIcq does not so slight preference to keep the former in its own file per convention and the latter in submit_tx . Wdyt?
> 
> So just to clarify, I meant something like this:
> 
> ```
> # msg_server_query_slashed_validator.go
> func QuerySlashedValidator () {
>     k.QueryValidatorExchangeRate()
> }
> 
> # msg_server_submit_tx.go
> func QueryValidatorExchangeRate() {
> 
> }
> func QueryValidatorShares() {
> 
> }
> ```
> 
> Idea being that QuerySlashedValidator is the user facing function. And QueryValidatorExchangeRate and QueryValidatorShares are the more backend functions that are triggered.
> 
> Is that how you interpreted my message?

I see, thanks for clarifying. yeah, that's a nice clean format and not a hard change to make. done.

sampocs commented:
beautiful - great job on this!

asalzmann commented on x/stakeibc/keeper/msg_server_update_validator_shares_exch_rate.go:
```diff
@@ -0,0 +1,15 @@
+	"github.com/Stride-Labs/stride/x/stakeibc/types"
+)
+
+// todo rename for clarity (this is the validator query : 1st step of daisy chain)
```
```suggestion
// This kicks off two ICQs, each with a callback, that will update the number of tokens on a validator
// after being slashed. The flow is
// 1. QueryValidatorExchangeRate
// 2. ValidatorCallback
// 3. QueryDelegationDaisyChain
// 4. DelegationCallback
```

asalzmann commented on x/stakeibc/keeper/callbacks.go:
```diff
@@ -166,3 +171,137 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
+				slashPct := sdk.NewDec(slashAmtInt64).Quo(sdk.NewDec(delAmtInt64))
+				k.Logger(ctx).Info(fmt.Sprintf("ICQ'd delAmt mismatch zone %s validator %s delegator %s records was %d icq shows %d slashAmt %d slashPct %d... UPDATING!",
+					zone.ChainId, v.Address, qdel.DelegatorAddress, v.DelegationAmt, qNumTokens, slashAmt, slashPct))
+				// TODO (not priority) move rate limiting logic to new rate limiting module
```
```suggestion
				// TODO(TEST-172): move rate limiting logic to new rate limiting module
```

asalzmann commented on x/stakeibc/keeper/callbacks.go:
```diff
@@ -166,3 +171,137 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
+			qNumTokens := qdel.Shares.Mul(v.TokensFromShares).TruncateInt()
+			k.Logger(ctx).Info(fmt.Sprintf("DelegationCallback: zone %s validator %s prevNtokens %v qNumTokens %v", zone.ChainId, v.Address, v.DelegationAmt, qNumTokens))
+			if qNumTokens.Uint64() < v.DelegationAmt {
+				// TODO(later) add some safety checks here (e.g. we could query the slashing module to confirm the decr in tokens was due to slash)
```
nit: add ticket

riley-stride replied:
done

asalzmann commented on x/stakeibc/keeper/callbacks.go:
```diff
@@ -166,3 +171,137 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
+				v.Weight = sdk.NewDec(weightInt64).Mul(weightMul).TruncateInt().Uint64()
+			}
+			// write back to state and break (reset TokensFromShares for clarity, so we're not tempted to use it again later)
+			v.TokensFromShares = sdk.NewDec(0)
```
Can we rename `TokensFromShares` -> `TokensFromSharesExchRate`? I was initially confused by this variable because it sounds like it's storing a number of tokens.

It feels a bit risky to only reset this to 0 at the end of the daisy-chained callback, I know that right now the only place we use this is in this code path, but if the daisy chain were to fail halfway through, the validator could have stale `TokensFromShares` across epochs. I see two possible ways to resolve this (curious if you can think of others)
1. At the beginning of each epoch, iterate all validators on all host zones and set `TokensFromShares` to 0 (slightly inefficient perhaps, but the safest approach)
2. Restructure this as follows
```
message ValidatorExchangeRate {
    string tokensFromSharesExchRate  = 1 [
        (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec",
        (gogoproto.nullable)   = false
    ];
    uint64 epochNumber = 2;

}
```
then add a function
```
func (k Keeper) GetInternalValidatorExchRate(v Validator) sdk.Dec {
    // check that the exchange rate on the validator is scoped to the current epoch
    // return it
}
```

riley-stride replied:
Scoping the internal exchange rate to an epochNumber (your suggestion #2) might solve this issue entirely. If we check the epochNumber on an internalExchRate before using it, why do we need to reset to 0 at the beginning the epoch (suggestion #1) ?

Will rename the variable, I agree it's confusing (borrowed from cosmos-sdk distr module nomenclature fwiw, but that doesn't mean we shouldn't improve it)

asalzmann replied:
I don't think we need to do both, either 1 or 2 should suffice

riley-stride replied:
followed approach 1, done

asalzmann commented on x/stakeibc/keeper/msg_server_submit_tx.go:
```diff
@@ -396,3 +365,91 @@ func (k Keeper) GetLightClientTimeSafely(ctx sdk.Context, connectionID string) (
 }
+
+// query and update validator exchange rate
+func (k Keeper) QueryValidatorExchangeRate(ctx sdk.Context, msg *types.MsgUpdateValidatorSharesExchRate) (*types.MsgUpdateValidatorSharesExchRateResponse, error) {
```
nit: feels like this should live in `x/stakeibc/keeper/msg_server_update_validator_shares_exch_rate.go`

riley-stride replied:
See discussion on this PR with Sam. I don't feel strongly between the two, but he made a good point re: the flow is more clear if both fns are in the same place

asalzmann commented on x/stakeibc/keeper/msg_server_submit_tx.go:
```diff
@@ -244,9 +249,13 @@ func (k Keeper) SubmitTxsEpoch(
+	bufferSizeParam := cast.ToInt64(k.GetParam(ctx, types.KeyBufferSize))
+	bufferSize := epochTracker.Duration / bufferSizeParam
+	timeoutNanos := epochTracker.NextEpochStartTime - bufferSize
+	// TODO safety: it's possible the cast below will handle this case with a graceful error, but leaving this here to make double sure until we can be sure
```
nit: rephrase for clarity

riley-stride replied:
replaced with ToInt64E cast

asalzmann commented on x/stakeibc/keeper/msg_server_submit_tx.go:
```diff
@@ -244,9 +249,13 @@ func (k Keeper) SubmitTxsEpoch(
+	bufferSize := epochTracker.Duration / bufferSizeParam
+	timeoutNanos := epochTracker.NextEpochStartTime - bufferSize
+	// TODO safety: it's possible the cast below will handle this case with a graceful error, but leaving this here to make double sure until we can be sure
+	if timeoutNanos < 0 {
```
timeoutNanos is absolute, not relative, so it shouldn't ever be l.t. zero

there's a check in the underlying `SendTx` that asserts that timeoutNanos is g.t. the current block time
```
if uint64(ctx.BlockTime().UnixNano()) >= timeoutTimestamp {
	return 0, icatypes.ErrInvalidTimeoutTimestamp
}
```

I think possible the check we want here is that bufferSize is less than one epoch and positive?

riley-stride replied:
timeoutNanos handled with To___E. 
Added the buffer check

asalzmann commented on x/stakeibc/keeper/keeper.go:
```diff
@@ -140,3 +143,36 @@ func (k Keeper) GetConnectionId(ctx sdk.Context, portId string) (string, error)
+		return sdk.ZeroDec(), sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "Failed to get epoch tracker for %s", epochType)
+	}
+	prevEpochStartTime := epochTracker.NextEpochStartTime - epochTracker.Duration
+	elapsedShare := sdk.NewDec(ctx.BlockTime().UnixNano() - prevEpochStartTime).Quo(sdk.NewDec(epochTracker.Duration))
```
nit: is it worth adding a check that `ctx.BlockTime().UnixNano() > prevEpochStartTime && ctx.BlockTime().UnixNano() < epochTracker.NextEpochStartTime`

riley-stride replied:
added this check and a check that `elapsedShare \in (0,1)`

asalzmann commented on x/stakeibc/keeper/keeper.go:
```diff
@@ -140,3 +143,36 @@ func (k Keeper) GetConnectionId(ctx sdk.Context, portId string) (string, error)
+		k.Logger(ctx).Error(fmt.Sprintf("Failed to get epoch tracker for %s", epochType))
+		return sdk.ZeroDec(), sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "Failed to get epoch tracker for %s", epochType)
+	}
+	prevEpochStartTime := epochTracker.NextEpochStartTime - epochTracker.Duration
```
```suggestion
	currEpochStartTime := epochTracker.NextEpochStartTime - epochTracker.Duration
```

asalzmann reviewed (changes requested):
Mostly minor comments, one more iteration and I think we should be good to go. The TTL thing is an issue that we'll need to solve before launch, I left a related comment around `TokensFromShares` (stale ICQ data stretching across multiple epochs) that I think we should resolve in this PR. 

nit: in PR description, update "(currently split 90/10)" -> "(currently split 80/20)"

riley-stride commented on x/interchainquery/types/keys.go:
```diff
@@ -20,6 +20,12 @@ const (
 	prefixQuery = iota + 1
 )
 
+// keys for proof queries to various stores
```
nice. added `go/onboarding`

asalzmann commented on x/stakeibc/keeper/callbacks.go:
```diff
@@ -176,3 +182,166 @@ func WithdrawalBalanceCallback(k Keeper, ctx sdk.Context, args []byte, query icq
+	// converting 1.0 gives us the exchange rate to later use in the next CB
+	v.InternalExchangeRate = &types.ValidatorExchangeRate{
+		InternalTokensToSharesRate: queriedValidator.TokensFromShares(sdk.NewDec(1.0)),
+		EpochNumber:                strideEpochTracker.GetEpochNumber(),
```
given we don't have TTLs, this is a bit misleading in that `EpochNumber` is "the epoch the query returned in", not the epoch it was issued in. just noting that the exchange rate might have been calculated at a previous `EpochNumber`

riley-stride replied:
> few minor comments, after resolving those good to merge I think. it does seem like there are a lot of commits towards the end of this PR, what are those? do you want a review on those as well?
> 
> we should prioritize adding TTLs, the edge cases are hard to reason about without them

TTLs are up next!

asalzmann reviewed (commented):
few minor comments, after resolving those good to merge I think.

we should prioritize adding TTLs, the edge cases are hard to reason about without them

asalzmann commented:
seems like there are a lot of commits towards the end of this PR, what are those?
